pep-0675.rst

```diff
@@ -913,6 +960,246 @@ is documentation, which is easily ignored and often not seen. With
 ``Literal[str]``, API misuse requires conscious thought and artifacts
 in the code that reviewers and future developers can notice.
 
+.. _appendix_C:
+
+Appendix C: ``str`` methods that preserve ``Literal[str]``
+==========================================================
+
+The ``str`` class has several methods that would benefit from
+``Literal[str]``. For example, users might expect
+``"hello".capitalize()`` to have the type ``Literal[str]`` similar to
+the other examples we have seen in the `Inferring Literal[str] section
+<inferring_literal_str>`_ section. Inferring the type ``Literal[str]``
+is correct because the string is not an arbitrary user-supplied
+string. In other words, the ``capitalize`` method preserves the
+``Literal[str]`` type. There are several other methods that preserve
+the ``Literal[str]``.
+
+We face a tradeoff in this PEP. We could require type checkers to
+preserve ``Literal[str]`` either (a) only for the four cases mentioned
+in the `Inferring Literal[str] section <inferring_literal_str_>`_
+section or (b) for all the ``str`` methods for which it would be
+valid. Option (a) might surprise users by losing the ``Literal[str]``
+type in innocuous uses, e.g., with ``my_literal.capitalize()``. Option
+(b) would be more user-friendly but would require some more work from
+type checkers.
+
+We decided to favor user-friendliness and go with option (b). However,
+if the Steering Council feels the other way, we are willing to go with
+option (a).
+
+Further, we propose updating the stub for ``str`` in typeshed so that
+the methods are overloads with the ``Literal[str]``-preserving
+versions. This means type checkers do not have hardcode
+``Literal[str]`` behavior for each method. This also lets us easily
+support new methods in the future by updating the typeshed stub.
+
+For example, to preserve literal types for the ``capitalize`` method,
+we would change the stub as below:
+
+::
+
+    # before
+    def capitalize(self) -> str: ...
+
+    # after
+    @overload
+    def capitalize(self: Literal[str]) -> Literal[str]: ...
+    @overload
+    def capitalize(self) -> str: ...
+
+The downside of changing ``str`` stub is that the stub becomes more
+complicated and can make error messages harder to understand. Type
+checkers may need to special-case ``str`` so that error messages are
+understandable for users.
+
+If the Steering Council is opposed to this typeshed stub change, we
```

Comment:
This feels too low-level for the SC to decide, unless you have concrete indications that it's something they're concerned about. I'd prefer if the PEP takes a concrete stance, otherwise it just becomes something that can drag out the discussion. (I don't want another PEP 646.)

Reply:
> I'd prefer if the PEP takes a concrete stance

Makes sense. I'll add the concrete stance we've agreed on and remove this offer.

> (I don't want another PEP 646.)

Too soon, Jelle. Too soon :)